From: =?utf-8?q?Guido_G=C3=BCnther?= <agx@sigxcpu.org>
Date: Sat, 20 Jul 2024 20:00:23 +0200
Subject: prompt: Only show when we got a request
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

So far we showed the prompt right away which led to some flicker when
the UI elements were made visible and can also lead to an empty dialog
when non of system_prompt_{password,confirm}_async are ever invoked.

Closes: https://gitlab.gnome.org/World/Phosh/phosh/-/issues/1092
Tested-By: Evangelos Ribeiro Tzaras <devrtz@fortysixandtwo.eu>
Signed-off-by: Guido Günther <agx@sigxcpu.org>

(cherry picked from commit 0ac1172757eadc09319331f8726a00f395baa9a1)
---
 src/system-prompt.c   | 11 +++++++++++
 src/system-prompter.c |  5 -----
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/src/system-prompt.c b/src/system-prompt.c
index 2ac22e1..05e0a86 100644
--- a/src/system-prompt.c
+++ b/src/system-prompt.c
@@ -13,6 +13,7 @@
 
 #include "phosh-config.h"
 
+#include "shell.h"
 #include "system-prompt.h"
 
 #define GCR_API_SUBJECT_TO_CHANGE
@@ -259,6 +260,11 @@ phosh_system_prompt_password_async (GcrPrompt *prompt,
   g_object_notify (obj, "warning-visible");
   g_object_notify (obj, "choice-visible");
   priv->shown = TRUE;
+
+  /* Show widget when not locked and keep that in sync */
+  g_object_bind_property (phosh_shell_get_default (), "locked",
+                          self, "visible",
+                          G_BINDING_SYNC_CREATE | G_BINDING_INVERT_BOOLEAN);
 }
 
 
@@ -318,6 +324,11 @@ phosh_system_prompt_confirm_async (GcrPrompt *prompt,
   g_object_notify (obj, "warning-visible");
   g_object_notify (obj, "choice-visible");
   priv->shown = TRUE;
+
+  /* Show widget when not locked and keep that in sync */
+  g_object_bind_property (phosh_shell_get_default (), "locked",
+                          self, "visible",
+                          G_BINDING_SYNC_CREATE | G_BINDING_INVERT_BOOLEAN);
 }
 
 
diff --git a/src/system-prompter.c b/src/system-prompter.c
index 8c87ef1..22fcfeb 100644
--- a/src/system-prompter.c
+++ b/src/system-prompter.c
@@ -40,11 +40,6 @@ new_prompt_cb (GcrSystemPrompter *prompter,
 
   prompt = phosh_system_prompt_new ();
 
-  /* Show widget when not locked and keep that in sync */
-  g_object_bind_property (phosh_shell_get_default (), "locked",
-                          prompt, "visible",
-                          G_BINDING_SYNC_CREATE | G_BINDING_INVERT_BOOLEAN);
-
   return GCR_PROMPT (prompt);
 }
 
